src/util/Spotify.js

```diff
@@ -48,53 +48,66 @@ export const Spotify = {
         Tracklist.js, cannot read property 'map' of undefined, and pointing on the
         traks of the TrackList.. I couldn't understand why.*/
     async search (term) {
-      /* let userAccessToken = this.getUserAccessToken();   I commented this out becaues it is already defined in the global of this file*/
+      
+      const topic = 'Searching Spotify\'s API.'
+
+      const mainTerm = encodeURI(term); // this converts the term passed as a parameter to a URI format
+
+      let response;
+      let JSONResponse;
+      let error;
+      let mappedTracks;
 
       if (!userAccessToken){
         console.log('No Access Token.');
-        return;
+        return ;
       };
-
-      const mainTerm = encodeURI(term); // this converts the term passed as a parameter to a URI format
-      
       
+      //The code below tries to contact Spotifies API
       try {
-        const response =  await fetch(`https://api.spotify.com/v1/search?type=track&q=${mainTerm}`, {
+        response =  await fetch(`https://api.spotify.com/v1/search?type=track&q=${mainTerm}`, {
         headers: {Authorization: `Bearer ${userAccessToken}`}
-      })
-      
-      const jsonifyResponse = async (response) => {
-
-        let responseOK = await response.ok;
+      })      
+      } catch (e) {
+        error = new Error(`Network error, while ${topic}.`);
+        return Promise.reject(error);
+      }
 
-          if(responseOK){
-              return response.json()
-          }
-          throw new Error('Request failed');
+      if (!response.ok) {
+        error = new Error(`Server error, status: ${response.status}, while ${topic}.`);
+        return Promise.reject(error);
       }
-      const jsonResponse = await jsonifyResponse(response);
 
-      let responseTracks = await jsonResponse.tracks;
-      
-      if (!responseTracks){
-        return [];
+      // The code below will try to convert the response to JSON format
+
+      try{
+        JSONResponse = await response.json();
+
+      } catch (err){
+        error = new Error(`Fail to read response, while ${topic}.`);
+        return Promise.reject(error);
       }
 
+      
+      let searchResults =  () => {
 
-      return jsonResponse.tracks.items.map(track => ({
+        if (!JSONResponse.tracks){
+          mappedTracks =  [];
+        }
+        mappedTracks = JSONResponse.tracks.items.map(track => ({
+
+          id: track.id,
+          name: track.name,
+          artist: track.artists[0].name,
+          album: track.album.name,

```

Comment:
just know that if you access a deep nested value like `track.album.name` and for some reason `album` does not exist you will hit a runtime error that will break everything 

in this case you should always check, especially if the data is coming from outside api.

this is a fallback to an empty string:

```
album : track.album ? track.album.name : ''
```

also, I highly recommend using lodash everywhere, it helps in such cases as well

checkout lodash `_.get`

Reply:
[https://lodash.com/docs/#get](url)

Reply:
looks like a super cool function. but where should I use it? in `Spotify.search` or wherever `Spotify.search` is called (in App.js for example)?

Reply:
you should use it whenever you access for a value in an object more then one level deep

Reply:
So in this specific case should I use it on the `JSONResponse` or only start using it inside the `.map` ?